lib/fizzy/execute.cpp

```diff
@@ -1145,42 +1143,41 @@ ExecutionResult execute(Instance& instance, FuncIdx func_idx, const Value* args,
         }
         case Instr::i64_div_s:
         {
-            auto const rhs = stack[0].as<int64_t>();
-            auto const lhs = stack[1].as<int64_t>();
+            const auto rhs = stack.pop().as<int64_t>();
+            const auto lhs = stack.top().as<int64_t>();
             if (rhs == 0 || (lhs == std::numeric_limits<int64_t>::min() && rhs == -1))
                 goto trap;
-            binary_op(stack, div<int64_t>);
+            stack.top() = div(lhs, rhs);
             break;
         }
         case Instr::i64_div_u:
         {
-            auto const rhs = stack.top().i64;
+            const auto rhs = stack.pop().i64;
             if (rhs == 0)
                 goto trap;
-            binary_op(stack, div<uint64_t>);
+            const auto lhs = stack.top().i64;
+            stack.top() = div(lhs, rhs);
             break;
         }
         case Instr::i64_rem_s:
         {
-            auto const rhs = stack.top().as<int64_t>();
+            const auto rhs = stack.pop().as<int64_t>();
             if (rhs == 0)
                 goto trap;
-            auto const lhs = stack[1].as<int64_t>();
+            const auto lhs = stack.top().as<int64_t>();
             if (lhs == std::numeric_limits<int64_t>::min() && rhs == -1)
-            {
-                stack.pop();
                 stack.top() = 0;
-            }
             else
-                binary_op(stack, rem<int64_t>);
+                stack.top() = rem(lhs, rhs);
```

Comment:
Instead of keeping logic here, alternative solution is to introduce `rem_s` and `rem_u` which does the wasm logic and keep using `binary_op`. This would be in-line with how bitops (like clz/ctz) work.

Reply:
Ah never mind, we kept these separate because they trap.

Reply:
> Other approach is to wrap division instruction with a bigger helpers returning bool for trap (as we have in other cases).

But I'd leave it for some other time because it is also worth to compare this version with that version.